File: apps/android/feature/exercises/domain/src/main/java/com/scribblefit/feature/exercises/domain/usecase/GetExerciseAIInsightUseCase.kt
package com.scribblefit.feature.exercises.domain.usecase

import com.scribblefit.core.model.AIInsight
import com.scribblefit.core.model.CurrentDate
import com.scribblefit.core.model.Exercise
import com.scribblefit.feature.ai.domain.LLMEngine
import com.scribblefit.feature.exercises.domain.ExerciseRepository
import java.time.Instant
import java.time.LocalDate
import java.time.ZoneId
import java.time.format.DateTimeFormatter

class GetExerciseAIInsightUseCase(
    private val llmEngine: LLMEngine,
    private val exerciseRepository: ExerciseRepository
) {
    private val dateFormatter = DateTimeFormatter.ofPattern("MMM d, yyyy")
        .withZone(ZoneId.systemDefault())

    suspend operator fun invoke(dateLookBehind: Long = 7): Result<AIInsight> {
        val currentDate = CurrentDate(LocalDate.now())
        val startDate = CurrentDate(LocalDate.now().minusDays(dateLookBehind))

        val history: List<Exercise> = exerciseRepository.getExercisesInRange(
            startDate = startDate.startOfDayInMillis,
            endDate = currentDate.startOfDayInMillis
        )

        if (history.isEmpty()) {
            return Result.failure(Exception("No history available to generate insights"))
        }

        val historyContext = history.joinToString("\n") { exercise ->
            val date = dateFormatter.format(Instant.ofEpochMilli(exercise.createdAt))
            val sets = exercise.sets.joinToString(", ") { "${it.weight}x${it.reps}" }
            "$date: $sets"
        }

        return llmEngine.generateExerciseInsight(historyContext)
    }
}

File: apps/ios/ScribbleFit/LocalPackages/Features/FeatureExercises/Sources/FeatureExercises/Domain/UseCases/GetExerciseAIInsightUseCase.swift
import Foundation
import CoreModel

@MainActor
public final class GetExerciseAIInsightUseCase {
    private let llmService: LLMService
    
    public init(llmService: LLMService) {
        self.llmService = llmService
    }
    
    public func execute(history: [ExerciseHistorySession]) async throws -> AIInsight {
        guard !history.isEmpty else {
            throw NSError(domain: "ExerciseAIInsight", code: 404, userInfo: [NSLocalizedDescriptionKey: "No history available to generate insights"])
        }
        
        let dateFormatter = DateFormatter()
        dateFormatter.dateStyle = .medium
        
        let historyContext = history.prefix(5).map { session in
            let date = dateFormatter.string(from: session.date)
            let sets = session.exercise.sets.map { "\($0.weight ?? 0)x\($0.reps)" }.joined(separator: ", ")
            return "\(date): \(sets)"
        }.joined(separator: "\n")
        
        return try await llmService.generateExerciseInsight(history: historyContext)
    }
}
